### PR TITLE
feature/predict-trades

### PR DESCRIPTION
**Pull request by arealclimber**
### DEVELOP

- [x]  lib/books/trade_book.ts
- [x] src/constants/code.ts
- [x] src/constants/model.ts

### CHECK LIST
- [x]  #626 
- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 0
- new Class / Component:  0
- new loop: 0
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/Global-Context-public-d93ac7c941c94fb6b173d34a7bc4e959)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/lib/books/trade_book.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    );
+
+    // Info: Average prices with same timestamp (20230522 - Shirley)
+    const timestampMap = new Map();
+    recentTrades.forEach(trade => {
+      if (timestampMap.has(trade.timestampMs)) {
+        const prev = timestampMap.get(trade.timestampMs);
+        timestampMap.set(trade.timestampMs, {
+          ...prev,
+          price: (prev.price + trade.price) / 2,
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:164]:
```
price: (prev.price * prev.quantity + trade.price * trade.quantity) / 2,

```

File: src/lib/books/trade_book.ts [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+    const prediction = newTrades.slice(-length);
+
+    return prediction;
+  }
+
+  private getLinearRegressionVariables(trades: ITrade[]): {m: number; b: number} {
+    // Info: (20230522 - Shirley) Step 2: Prepare data for regression
+    const x = trades.map((t, i) => t.timestampMs);
+    const y = trades.map(t => t.price);
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:205]:
```tsx
const y = trades.map(t => t.price * t.quantity);


```

File: src/lib/books/trade_book.ts [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+const TradeBookInstance = new TradeBook({
+  intervalMs: 100,
+  model: Model.LINEAR_REGRESSION,
+  minLengthForLinearRegression: 2,
+  minMsForLinearRegression: 1000 * 30,
+});
+
+export default TradeBookInstance;
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:331]: export default Class TradeBook

**Review by Luphia — APPROVED** [review]
- no bug found

**Pull request merged**